meadow-client/src/MainFrame.purs

```diff
@@ -595,18 +605,23 @@ evalF (ResetSimulator next) = do
   resetContract
   pure next
 
---  mContents <- withMarloweEditor Editor.getValue
---  case mContents of
---    Nothing -> pure next
---    Just contents -> do
---      let contract = parse Parser.contract contents
---      case contract of
---        Right c -> do
---          assign _marloweCompileResult $ Left [MarloweError "oh no"]
---          pure next
---        Left err -> do
---          assign _marloweCompileResult $ Left [MarloweError err]
---          pure next
+evalF (Undo next) = do
+  modifying (_marloweState) removeState
+  mCurrContract <- use (_marloweState <<< _Head <<< _contract)
+  case mCurrContract of
+    Just currContract -> do
+      void $ withMarloweEditor $ Editor.setValue (show $ pretty currContract) (Just 1)
+      pure next
+    Nothing -> pure next
+  where
+  removeState ms =
+    let
+      {head, tail} = NEL.uncons ms
+    in
+      case NEL.fromList tail of
+        Nothing -> ms
+        Just netail -> netail
+
```

Comment:
I would be tempted to call reset once there is no more transactions to undo, but maybe we should discuss and have a separate issue for this.

Reply:
With the disabling of the Undo button this doesn't feel necessary. When I used it, it felt clear to me, the Undo button is disabled, there is possibly an item in the input composer and the reset button is available. It feels quite intuitive (however I am likely quite biased since I've been using it a lot).